prow/resultstore/writer/writer.go

```diff
@@ -117,26 +127,69 @@ func New(ctx context.Context, log *logrus.Entry, client ResultStoreBatchClient,
 	}
 	ctx, cancel := context.WithTimeout(ctx, rpcRetryDuration)
 	defer cancel()
-	err := wait.ExponentialBackoffWithContext(ctx, rpcRetryBackoff, func() (bool, error) {
+
+	err := w.createInvocation(ctx, inv, invID)
+	if err == nil {
+		return w, nil
+	}
+	if !IsAlreadyExistsErr(err) {
+		return nil, err
+	}
+
+	if touchErr := w.touchInvocation(ctx, invID); PermanentError(touchErr) {
+		// Since it was confirmed above that the Invocation exists, a
+		// permanent error here indicates the Invocation is finalized.
+		return nil, err
+	}
+
+	if err = w.retrieveResumeToken(ctx, invID); err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+func (w *writer) createInvocation(ctx context.Context, inv *resultstore.Invocation, invID string) error {
```

Comment:
Previously we had just 1 function which was retried, but now we have 3 (createInvocation, touchInvocation, retrieveResumeToken). Without reading too deeply, was there a reason why you could not just use the 1 retry function?

Reply:
As a principle, serial RPCs must be retried individually to ensure reasonable progress during high error rates.

For example, with a 90% error rate, there is only a 0.1% probability of success across three calls. In this scenario, 1000 retries would be required on average, so the client presents much more load than necessary on the likely overloaded server. With independent retries, the code doesn't reissue any previously successful calls, and a total of only 30 retries would be needed on average.

Reply:
Makes sense. LGTM, but I wonder if there's a way to reduce the code duplication across these 3. Seems a bit un-ergonomic right now.

Reply:
I've added a third commit to simply the error filtering which seems to be the repetitious part of these functions. The resulting functions are largely unique.